HTML5-frontend/ndt-library.js

```diff
@@ -0,0 +1,234 @@
+/**
+ * @fileoverview This javascript code includes an NDT object that contains the
+ * methods that a customized front end implemetation of the NDT test could
+ * utilize.  The NDT object is self invoked when the library is included on a
+ * page, however, the frontend can customize certain properties when the
+ * interaction is initialized with the startTest method.
+ *
+ * Dependencies: ndt-wrapper.js, ndt-browser-client.js
+ */
+
+
+/**
+ * Immediately invoked function expression that creates NDT library object
+ */
+(function (window) {
+  'use strict';
+
+  /** @constructor */
+  function NDTLibjs() {
+
+    // Look up a suitable NDT server using mlab-ns. The ndtServer property is
+    //  used by the backend javascript to determine which server to test
+    //  against.
+    this.getServer();
+
+    // Initializes simulate property to false.  Can be changed upon starting a
+    // test in the testServer method
+    this.simulate = false;
+
+    // STATUS VARIABLES
+    this.use_websocket_client = false;
+    this.websocket_client = null;
+
+  }
+
+  /**
+   * Operates on an instance of NDT to begin executing a test.
+   * @param - none
+   * @return - {string} True if client is using websockets, False if error
+   * occurred during finding backend server, or applet container element fall
+   * back to Java applet.
+   */
+  NDTLibjs.prototype.startTest = function() {
+    if (NDT.ndtServer) {
+      this.checkInstalledPlugins();
+      var clientProtocol = this.createBackend();
+      this.testNDT().run_test(this.ndtServer);
+    } else {
+      // in case AJAX call fails or is taking longer
+      this.getServer();
+      return false;
+    }
+    return clientProtocol;
+  }
+
+  /**
+   * Operates on an instance of NDT to determine if the browser supports use of
+   * websockets.  If not, will attempt to use Java if available.
+   */
+  NDTLibjs.prototype.checkInstalledPlugins = function() {
+    var hasJava = false;
+    var hasWebsockets = false;
+
+    hasJava = true;
+    if (typeof deployJava !== 'undefined') {
+      if (deployJava.getJREs() == '') {
+        hasJava = false;
+      }
+    }
+    hasWebsockets = false;
+    try {
+      var ndt_js = new NDTjs();
+      if (ndt_js.checkBrowserSupport()) {
+        hasWebsockets = true;
+      }
+    } catch(e) {
+      hasWebsockets = false;
+    }
+
+    if (hasWebsockets) {
+      this.use_websocket_client = true;
+
+    }
+    else if (hasJava) {
+      this.use_websocket_client = false;
+
+    }
+  }
+
+  /**
+   * Creates the backend for the NDT object.  Tries to create websocket backend
+   * if supported, falls back to java applet if websocket not available
+   * @return - {obj} If Java is to be used returns Java applet information,
+   *     otherwise returns false
+   */
+  NDTLibjs.prototype.createBackend = function() {
+    if (this.use_websocket_client) {
```

Comment:
I would opt for these functions to be populated with variables like use_websocket_client passed as parameters, rather that use instance properties